In [1]:
import pandas as pd
import numpy as np
import seaborn as sns

In [2]:
books=pd.read_csv(r"D:\Data Science\Python\ML_final_projects\Books Reccomandation\Books.csv")

C:\Users\hirap\AppData\Local\Temp\ipykernel_5252\1079732174.py:1: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  books=pd.read_csv(r"D:\Data Science\Python\ML_final_projects\Books Reccomandation\Books.csv")


In [3]:
ratings=pd.read_csv(r"D:\Data Science\Python\ML_final_projects\Books Reccomandation\Ratings.csv")

In [4]:
users=pd.read_csv(r"D:\Data Science\Python\ML_final_projects\Books Reccomandation\Users.csv")

In [5]:
books.head()

ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author Year-Of-Publication                   Publisher  \
0    Mark P. O. Morford                2002     Oxford University Press   
1  Richard Bruce Wright                2001       HarperFlamingo Canada   
2          Carlo D'Este                1991             HarperPerennial   
3      Gina Bari Kolata                1999        Farrar Straus Giroux   
4       E. J. W. Barber                1999  W. W. Norton &amp; Company   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

In [6]:
ratings.head()

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5
2   276727  0446520802            0
3   276729  052165615X            3
4   276729  0521795028            6

In [7]:
users.head()

User-ID                            Location   Age
0        1                  nyc, new york, usa   NaN
1        2           stockton, california, usa  18.0
2        3     moscow, yukon territory, russia   NaN
3        4           porto, v.n.gaia, portugal  17.0
4        5  farnborough, hants, united kingdom   NaN

In [8]:
ratings['Book-Rating'].value_counts() #rating is between 0 to 10

Book-Rating
0     716109
8     103736
10     78610
7      76457
9      67541
5      50974
6      36924
4       8904
3       5996
2       2759
1       1770
Name: count, dtype: int64

Checking the unique books count

In [9]:
books['ISBN'].nunique()

271360

In [10]:
ratings['ISBN'].nunique()

340556

# Report

In books dataset, there are 271,360 unique books. However, the rating dataset contains additional books not in my dataset, so we can only analyze the 271,360 books available to us.


We only require user IDs for our analysis, and the rating dataset contains this information. We can disregard user age and location details, which are present in the rating dataset but are not needed for our current analysis.

In [11]:
ratings['User-ID'].nunique()

105283

In [12]:
books['Book-Title'].nunique()

242135

In [13]:
books=pd.merge(books,ratings,on=["ISBN"])      #Merging the ratings and books dataset

In [14]:
books.head()

ISBN           Book-Title           Book-Author Year-Of-Publication  \
0  0195153448  Classical Mythology    Mark P. O. Morford                2002   
1  0002005018         Clara Callan  Richard Bruce Wright                2001   
2  0002005018         Clara Callan  Richard Bruce Wright                2001   
3  0002005018         Clara Callan  Richard Bruce Wright                2001   
4  0002005018         Clara Callan  Richard Bruce Wright                2001   

                 Publisher                                        Image-URL-S  \
0  Oxford University Press  http://images.amazon.com/images/P/0195153448.0...   
1    HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...   
2    HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...   
3    HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...   
4    HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0002005018.0...   
3  http://images.amazon.com/images/P/0002005018.0...   
4  http://images.amazon.com/images/P/0002005018.0...   

                                         Image-URL-L  User-ID  Book-Rating  
0  http://images.amazon.com/images/P/0195153448.0...        2            0  
1  http://images.amazon.com/images/P/0002005018.0...        8            5  
2  http://images.amazon.com/images/P/0002005018.0...    11400            0  
3  http://images.amazon.com/images/P/0002005018.0...    11676            8  
4  http://images.amazon.com/images/P/0002005018.0...    41385            0

In [15]:
books['Book-Title'].nunique()

241071

In [16]:
books['ISBN'].nunique()

270151

# Analysis
So from this analysis we can say that unique ISBN is high as compared to Book-Title. because some of book-titles are same, but the length of the dataset is nearly 10 Lac.

In [17]:
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1031136 entries, 0 to 1031135
Data columns (total 10 columns):
 #   Column               Non-Null Count    Dtype 
---  ------               --------------    ----- 
 0   ISBN                 1031136 non-null  object
 1   Book-Title           1031136 non-null  object
 2   Book-Author          1031134 non-null  object
 3   Year-Of-Publication  1031136 non-null  object
 4   Publisher            1031134 non-null  object
 5   Image-URL-S          1031136 non-null  object
 6   Image-URL-M          1031136 non-null  object
 7   Image-URL-L          1031132 non-null  object
 8   User-ID              1031136 non-null  int64 
 9   Book-Rating          1031136 non-null  int64 
dtypes: int64(2), object(8)
memory usage: 78.7+ MB


In [18]:
books['Year-Of-Publication'] = pd.to_numeric(books['Year-Of-Publication'], errors='coerce')
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1031136 entries, 0 to 1031135
Data columns (total 10 columns):
 #   Column               Non-Null Count    Dtype  
---  ------               --------------    -----  
 0   ISBN                 1031136 non-null  object 
 1   Book-Title           1031136 non-null  object 
 2   Book-Author          1031134 non-null  object 
 3   Year-Of-Publication  1031132 non-null  float64
 4   Publisher            1031134 non-null  object 
 5   Image-URL-S          1031136 non-null  object 
 6   Image-URL-M          1031136 non-null  object 
 7   Image-URL-L          1031132 non-null  object 
 8   User-ID              1031136 non-null  int64  
 9   Book-Rating          1031136 non-null  int64  
dtypes: float64(1), int64(2), object(7)
memory usage: 78.7+ MB


In [19]:
books.isnull().sum()

ISBN                   0
Book-Title             0
Book-Author            2
Year-Of-Publication    4
Publisher              2
Image-URL-S            0
Image-URL-M            0
Image-URL-L            4
User-ID                0
Book-Rating            0
dtype: int64

In [20]:
books.dropna(inplace=True)
books.isnull().sum()

ISBN                   0
Book-Title             0
Book-Author            0
Year-Of-Publication    0
Publisher              0
Image-URL-S            0
Image-URL-M            0
Image-URL-L            0
User-ID                0
Book-Rating            0
dtype: int64

In [21]:
books.duplicated().sum()

0

# Popularity Based Reccomandation System

we are selecting only those books, on which atleast 200 users give the rating
and those books has highest average rating.

In [22]:
ratings_with_name= books.copy()

In [23]:
num_rating_df=ratings_with_name.groupby('Book-Title').count()['Book-Rating'].reset_index()
num_rating_df.rename(columns={'Book-Rating':'num_ratings'},inplace=True)

Now we will do average of rating

In [24]:
avg_rating_df = ratings_with_name.groupby('Book-Title')['Book-Rating'].mean().reset_index()
avg_rating_df.rename(columns={'Book-Rating':'avg_ratings'}, inplace=True)

In [25]:
avg_rating_df

Book-Title  avg_ratings
0        A Light in the Storm: The Civil War Diary of ...     2.250000
1                                   Always Have Popsicles     0.000000
2                    Apple Magic (The Collector's series)     0.000000
3        Ask Lily (Young Women of Faith: Lily Series, ...     8.000000
4        Beyond IBM: Leadership Marketing and Finance ...     0.000000
...                                                   ...          ...
241060                                      Ã?Â?lpiraten.     0.000000
241061                     Ã?Â?rger mit Produkt X. Roman.     5.250000
241062                                Ã?Â?sterlich leben.     7.000000
241063                              Ã?Â?stlich der Berge.     2.666667
241064                                  Ã?Â?thique en toc     4.000000

[241065 rows x 2 columns]

In [26]:
popularity_df = num_rating_df.merge(avg_rating_df, on=['Book-Title'])
popularity_df

Book-Title  num_ratings  \
0        A Light in the Storm: The Civil War Diary of ...            4   
1                                   Always Have Popsicles            1   
2                    Apple Magic (The Collector's series)            1   
3        Ask Lily (Young Women of Faith: Lily Series, ...            1   
4        Beyond IBM: Leadership Marketing and Finance ...            1   
...                                                   ...          ...   
241060                                      Ã?Â?lpiraten.            2   
241061                     Ã?Â?rger mit Produkt X. Roman.            4   
241062                                Ã?Â?sterlich leben.            1   
241063                              Ã?Â?stlich der Berge.            3   
241064                                  Ã?Â?thique en toc            2   

        avg_ratings  
0          2.250000  
1          0.000000  
2          0.000000  
3          8.000000  
4          0.000000  
...             ...  
241060     0.000000  
241061     5.250000  
241062     7.000000  
241063     2.666667  
241064     4.000000  

[241065 rows x 3 columns]

In [27]:
popularity_df = popularity_df[popularity_df['num_ratings']>200].sort_values('avg_ratings',ascending=False).head(75)

I need books author and image so we will merge the dataframes

In [28]:
popularity_df=popularity_df.merge(books,on=['Book-Title']).drop_duplicates('Book-Title')[['Book-Title','Book-Author','Publisher','Image-URL-M','num_ratings','avg_ratings']]

# Collaborative Filtering Based Recommender System

"Identify users who have rated a book a maximum of 200 times and books that have received more than 40 ratings from those users."

In [29]:
x = ratings_with_name.groupby('User-ID').count()['Book-Rating'] >200
rating_given_users=x[x].index

In [30]:
filtered_ratings=ratings_with_name[ratings_with_name['User-ID'].isin(rating_given_users)]

In [43]:
y= filtered_ratings.groupby('Book-Title').count()['Book-Rating'] >=40

In [44]:
popular_books = y[y].index

In [45]:
final_ratings=filtered_ratings[filtered_ratings['Book-Title'].isin(popular_books)]

In [46]:
final_ratings.drop_duplicates()  # there is no duplicate value in my dataframe

ISBN                       Book-Title      Book-Author  \
31       0399135782           The Kitchen God's Wife          Amy Tan   
33       0399135782           The Kitchen God's Wife          Amy Tan   
34       0399135782           The Kitchen God's Wife          Amy Tan   
38       0399135782           The Kitchen God's Wife          Amy Tan   
39       0399135782           The Kitchen God's Wife          Amy Tan   
...             ...                              ...              ...   
1028600  0394429869  I Know Why the Caged Bird Sings     Maya Angelou   
1028602  0449001164                      The Promise      CHAIM POTOK   
1028815  0743527631         The Pillars of the Earth      Ken Follett   
1028817  0745168086              The Handmaid's Tale  Margaret Atwood   
1030506  0345478940                      Angel Falls   KRISTIN HANNAH   

         Year-Of-Publication            Publisher  \
31                    1991.0     Putnam Pub Group   
33                    1991.0     Putnam Pub Group   
34                    1991.0     Putnam Pub Group   
38                    1991.0     Putnam Pub Group   
39                    1991.0     Putnam Pub Group   
...                      ...                  ...   
1028600               1996.0         Random House   
1028602               1997.0     Ballantine Books   
1028815               2002.0               Encore   
1028817               1999.0  Chivers Audio Books   
1030506               2004.0     Ballantine Books   

                                               Image-URL-S  \
31       http://images.amazon.com/images/P/0399135782.0...   
33       http://images.amazon.com/images/P/0399135782.0...   
34       http://images.amazon.com/images/P/0399135782.0...   
38       http://images.amazon.com/images/P/0399135782.0...   
39       http://images.amazon.com/images/P/0399135782.0...   
...                                                    ...   
1028600  http://images.amazon.com/images/P/0394429869.0...   
1028602  http://images.amazon.com/images/P/0449001164.0...   
1028815  http://images.amazon.com/images/P/0743527631.0...   
1028817  http://images.amazon.com/images/P/0745168086.0...   
1030506  http://images.amazon.com/images/P/0345478940.0...   

                                               Image-URL-M  \
31       http://images.amazon.com/images/P/0399135782.0...   
33       http://images.amazon.com/images/P/0399135782.0...   
34       http://images.amazon.com/images/P/0399135782.0...   
38       http://images.amazon.com/images/P/0399135782.0...   
39       http://images.amazon.com/images/P/0399135782.0...   
...                                                    ...   
1028600  http://images.amazon.com/images/P/0394429869.0...   
1028602  http://images.amazon.com/images/P/0449001164.0...   
1028815  http://images.amazon.com/images/P/0743527631.0...   
1028817  http://images.amazon.com/images/P/0745168086.0...   
1030506  http://images.amazon.com/images/P/0345478940.0...   

                                               Image-URL-L  User-ID  \
31       http://images.amazon.com/images/P/0399135782.0...    11676   
33       http://images.amazon.com/images/P/0399135782.0...    36836   
34       http://images.amazon.com/images/P/0399135782.0...    46398   
38       http://images.amazon.com/images/P/0399135782.0...   113270   
39       http://images.amazon.com/images/P/0399135782.0...   113519   
...                                                    ...      ...   
1028600  http://images.amazon.com/images/P/0394429869.0...   239594   
1028602  http://images.amazon.com/images/P/0449001164.0...   239594   
1028815  http://images.amazon.com/images/P/0743527631.0...   240144   
1028817  http://images.amazon.com/images/P/0745168086.0...   240144   
1030506  http://images.amazon.com/images/P/0345478940.0...   244688   

         Book-Rating  
31                 9  
33                 0  
34                 9  
38                 0  
39                 0  
...              ...

In [47]:
pt=final_ratings.pivot_table(index='Book-Title',columns='User-ID',values='Book-Rating')

In [49]:
pt.fillna(0,inplace=True)

#### Now we import the cosine similarity 

In [52]:
from sklearn.metrics.pairwise import cosine_similarity

In [60]:
simliarity_score = cosine_similarity(pt)

In [74]:
def recommend(book_name):
    #index fetch
    index=np.where(pt.index==book_name)[0][0]
    similar_book = sorted(list(enumerate(simliarity_score[index])),key= lambda x:x[1],reverse=True)[1:6]

    for i in similar_book:
        print(pt.index[i[0]])
    

In [77]:
recommend('1984')

Animal Farm
The Handmaid's Tale
Awakening
Slaughterhouse Five or the Children's Crusade: A Duty Dance With Death
Brave New World
